### PR TITLE
Update readme

### PR DESCRIPTION
**Pull request by jtsiddons**
Noticed the README was out-of-date and the function calls needed to be updated/corrected.

I also changed the pip install instructions to the github path, since `cdm_reader_mapper` is not on PyPI yet.

### CODE REVIEW

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Not sure about the cdm model name in the mapper section, is this a valid model name?

**Comment by ludwiglierhammer** [comment]
@jtsiddons, first of all, thanks for updating the ``README``.

> @ludwiglierhammer: Not sure about the cdm model name in the mapper section, is this a valid model name?

 ``icoads_r3000_d701`` is no longer a valid cdm model name.
  The new structure for both ``mdf_reader`` and ``cdm_mapper`` is the following:

- data_model: e.g. ``icoads``
- release: e.g. ``r300`` or ``302``; I don't know why we had 4 zeros before; the original release name is ``Release 3.0.0`` or ``Release 3.0.2``
- deck: e.g. ``d701``

This information is combinde to ``icoads_r300_d701``. This name is split into a list of entries: 

-> ["icoads", "r300", "d701"].

New tables are searched for each entry and the previous ones are updated. If there are no tables, the previous ones are not updated. In our example:

- We start with the first entry ``icoads`` and read in the [ICOADS tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads).
- The second entry is the release. The previous ICOADS tables are updated with the [release tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads/r300).
- Finally, we use the [deck tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads/r300/d701) for some updates.

In the case of ``icoads_r300_d703``, there are no deck tables. The creation of the mapping tables ends with the release stage without any error messages. Instead, you get a warning that there are no deck tables. So for ICOADS it is the easiest way to name the cdm model in this structure. As a user you don't have to think about wich tables are needed for which deck and so on.

I hope this information is helpful.

By the way, could you please adjust the names in the notebooks? I recognized that some names are no longer valid:
- [cell 9](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/CDM_mapper_example_deck704.ipynb): ``name_of_model = "icoads_r3000_d704"`` to ``name_of_model = "icoads_r300_d704"``; this is important because ``r3000`` is not available and therfore ``r3000_d704`` too. Only IOCADS mapping tables are used although there are both ``r300`` and ``d704`` tables for updating
- [cell 2](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/CLIWOC_datamodel.ipynb): ``schema = "icoads_r300"`` to ``schema = "icoads_r300_d714"``; this doesn't effect the table creation, but would look more uniform I think.
- [cell 3](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): ``data_raw = mdf_reader.read(data.get("source"), imodel="icoads")`` to ``data_raw = mdf_reader.read(data.get("source"), imodel="icoads_r300_d701")``. This is similar to the first point.
- [cell 5](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): do you just want a copy of the ICOADS tables without any release or deck updates? 
- [cell 6](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): ``my_model_name = "imma1_d701"`` to ``my_model_name = "imma1_r300_d701"``; for consistency reasons only.

Do you agree with those changes?

File: README.rst [hunk 1]
```diff
@@ -78,15 +78,16 @@ This will set the file :code:`log_file.log` as the output for all logging inform
...
 
-    import cdm_reader_mapper as cdm
+    from cdm_reader_mapper.mdf_reader import read
+    from cdm_reader_mapper.data import test_data
 
-    data = cdm.tests.read_imma1_buoys_nosupp()
+    data = test_data.test_icoads_r300_d701.get("source")
 
-    imma_data = cdm.read(filepath, data_model="imma1", sections=["core", "c1", "c98"])
+    imma_data = read(filepath, imodel="icoads", sections=["core", "c1", "c98"])
```
Comment (ludwiglierhammer) [inline README.rst]:
```suggestion
    imma_data = read(filepath, imodel="icoads_r300_d701", sections=["core", "c1", "c98"])
```

File: README.rst [hunk 2]
```diff
@@ -95,11 +96,13 @@ Map this data to a CDM build for the same deck (in this case deck 704: US Marine
 
 .. code-block:: python
 
+    from cdm_reader_mapper.cdm_mapper import map_model
+
     name_of_model = "icoads_r3000_d704"
```
Comment (ludwiglierhammer) [inline README.rst]:
```suggestion
    name_of_model = "icoads_r300_d704"
```

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Thank you for this - apologies for not getting back sooner. I've added my thoughts to each of your points. I think I should expand on some of the notebooks to clarify why certain schema are being used (for example using a more generic schema for data which already has a full schema ("icoads" vs "icoads_r300_d701")).

> @jtsiddons, first of all, thanks for updating the `README`.
> 
> > @ludwiglierhammer: Not sure about the cdm model name in the mapper section, is this a valid model name?
> 
> `icoads_r3000_d701` is no longer a valid cdm model name. The new structure for both `mdf_reader` and `cdm_mapper` is the following:
> 
>     * data_model: e.g. `icoads`
> 
>     * release: e.g. `r300` or `302`; I don't know why we had 4 zeros before; the original release name is `Release 3.0.0` or `Release 3.0.2`
> 
>     * deck: e.g. `d701`
> 
> 
> This information is combinde to `icoads_r300_d701`. This name is split into a list of entries:
> 
> -> ["icoads", "r300", "d701"].
> 
> New tables are searched for each entry and the previous ones are updated. If there are no tables, the previous ones are not updated. In our example:
> 
>     * We start with the first entry `icoads` and read in the [ICOADS tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads).
> 
>     * The second entry is the release. The previous ICOADS tables are updated with the [release tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads/r300).
> 
>     * Finally, we use the [deck tables](https://github.com/glamod/cdm_reader_mapper/tree/main/cdm_reader_mapper/cdm_mapper/tables/icoads/r300/d701) for some updates.
> 
> 
> In the case of `icoads_r300_d703`, there are no deck tables. The creation of the mapping tables ends with the release stage without any error messages. Instead, you get a warning that there are no deck tables. So for ICOADS it is the easiest way to name the cdm model in this structure. As a user you don't have to think about wich tables are needed for which deck and so on.
> 
> I hope this information is helpful.

👍 Very helpful - clarifies the changes for me.

I'll make the changes in the README - I also think I should remove the `@v0.3.0` tag in the install instructions, as the API differs and the README instructions won't apply to that version. Could potentially set it to `v0.4.0` or the tag for the next released version.

> By the way, could you please adjust the names in the notebooks? I recognized that some names are no longer valid:
> 
>     * [cell 9](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/CDM_mapper_example_deck704.ipynb): `name_of_model = "icoads_r3000_d704"` to `name_of_model = "icoads_r300_d704"`; this is important because `r3000` is not available and therfore `r3000_d704` too. Only IOCADS mapping tables are used although there are both `r300` and `d704` tables for updating

I'll fix this.
 
>     * [cell 2](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/CLIWOC_datamodel.ipynb): `schema = "icoads_r300"` to `schema = "icoads_r300_d714"`; this doesn't effect the table creation, but would look more uniform I think.

This is fine - although there appears to be some more discrepencies in this notebook. The first read seems like it should be deck 714, then it uses the 703 schema to create a 730 schema. I think this was an original notebook so I look into the notebook. I'll also re-factor to just create a copy of the json file for the schema rather than use `read_schema`.
 
>     * [cell 3](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): `data_raw = mdf_reader.read(data.get("source"), imodel="icoads")` to `data_raw = mdf_reader.read(data.get("source"), imodel="icoads_r300_d701")`. This is similar to the first point.

In this case I do not want to use the specific deck schema - since the idea of the notebook is to show how one could build the c99 components. So in this case I think I need the generic "icoads" schema.

I'll add some extra text explaining why in this example we use the "icoads" schema. I could later read using the correct schema and compare the outputs (for the subset of the components the custom schema reads!).
 
>     * [cell 5](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): do you just want a copy of the ICOADS tables without any release or deck updates?

The idea behind this notebook was that we have a generic imma file that contains a "c99" section that doesn't apply to a known deck (or one that has an existing schema for c99). 

I want to add c99 information to the generic "icoads" schema. I use the `read_schema` function to essentially copy the "icoads" schema. In the updates `read_schema` now adds `PosixPath`s to the resulting dictionary (pointing to the file paths for required by the reader), so I need to remove these.

Maybe a better approach would to create a copy of the schema (in the temporary dir created for the example), read that copy in with `json.load`, add the new c99 fields (since I want to show how adding new fields can be done), then save it back to that copy.

I also require a copy of the "icoads" code tables, I don't want the specific deck 701 tables - maybe I should create one of them in the notebook as this would be a component of building a schema if the field is of type `key`.

>     * [cell 6](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb): `my_model_name = "imma1_d701"` to `my_model_name = "imma1_r300_d701"`; for consistency reasons only.

Happy to rename - since this is intended as a custom schema, I think that it should be easily distinguishable from the internal schema when reading the notebook. Maybe something along the lines of: "my_imma_v1_d701", or even "my_model" could be effective here?

File: README.rst (hunk 1 above)
Comment (jtsiddons) [reply README.rst]: See 9026d6c

File: README.rst (hunk 2 above)
Comment (jtsiddons) [reply README.rst]: See 9026d6c

**Comment by ludwiglierhammer** [comment]
@jtsiddons: Thanks or your comments and your efforts.

In case of reading external schema or code table files you can make a copy of the original ones, read it with ``json.load`` and do your manipulations. After writing the files on disk, you can simply use the the parameters in ``mdf_reader.read`` for reading external files:

1. set external schema path. **Note:** The schema file's name has to be the last directory in the external schema path plus the suffix ``.json``. 
E.g. ext_schema_path=``/path/to/my/schema/icoads`` -> schema file's name = ``icoads.json``

```
from cdm_reader_mapper import mdf_reader

read_ = mdf_reader.read(
    ...,
   ext_schema_path=<path_to_your_schema_file>,
)
``` 

2. Set external schema file directly.

```
from cdm_reader_mapper import mdf_reader

read_ = mdf_reader.read(
    ...,
   ext_schema_file=<path_to_your_schema_file_plus_filename>,
)
```

3. Set external code table path. Copy all needed code tables into this folder.

```
from cdm_reader_mapper import mdf_reader

read_ = mdf_reader.read(
    ...,
   ext_table_path=<path_to_your_schema_file>,
)
```
See also issue https://github.com/glamod/cdm_reader_mapper/issues/47.

With those new arguments you don't need the internal functions ``read_schema`` and ``read_table``.

File: README.rst [hunk 3]
```diff
@@ -95,11 +98,13 @@ Map this data to a CDM build for the same deck (in this case deck 704: US Marine
...
-    name_of_model = "icoads_r3000_d704"
+    from cdm_reader_mapper.cdm_mapper import map_model
 
-    cdm_dict = cdm.map_model(
-        name_of_model,
+    name_of_model = "icoads_r300_d704"
+
+    cdm_dict = map_model(
         data_raw,
+        name_of_model,
```
Comment (ludwiglierhammer) [inline README.rst]:
```suggestion
        imodel=name_of_model,
```

Comment (ludwiglierhammer) [reply README.rst]: It looks nicer somehow :blush:, isn't it?

Comment (ludwiglierhammer) [reply README.rst]: To take over this suggestion you can simply use the ``Commit suggestion`` button.

**Comment by jtsiddons** [comment]
@ludwiglierhammer: Thanks for the help/suggestions.

I'm still looking at the other two notebooks with model construction.

For [create_data_model.ipynb](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/create_data_model.ipynb), I am going to copy the schema json file directly, as it builds on the just the generic icoads schema. I've done this, I'm just thinking about/cleaning up the text sections to make sure it is clear/clean.

For [CLIWOC_datamodel.ipynb](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/example_notebooks/CLIWOC_datamodel.ipynb) this is slightly more complicated as the original version just copied the "idoads_r300_d730" schema. Following the changes the file in this case is _just_ c99 (which I agree is an improvement - ensuring consistency between common sections). I think this notebook would be a better showcase of the differences between reading with just the "icoads" schema and reading with the schema that includes c99 sections for d730. What do you think?

**Comment by ludwiglierhammer** [comment]
Aren't the two notebooks almost identical? We create a new schema file and new code tables in both notebooks. I think we could perhaps combine them into one. Do you agree with that? I don't quite understand what the additional benefit of the CLIWOC notebook is.

**Comment by jtsiddons** [comment]
They are certainly similar. I think the CLIWOC example is perhaps a more useful example case.

I can combine them. I'll refactor the CLIWOC notebook to include the manual construction of a subset of the c99 sections (similar to "create_data_model.ipynb"), and end by comparing with the in-built schema.

I'll then remove the "create_data_model" notebook.

I'll get that sorted first thing in the morning, if you're happy with that.

**Comment by ludwiglierhammer** [comment]
I absolutely agree with that.

After removing the "create_data_model" notebook, please adjust the last line in the "mdf_reader_test_overview" notebook and delete line 53 (example_notebooks/create_data_model.ipynb) in the [documentation index file](https://github.com/glamod/cdm_reader_mapper/blob/main/docs/index.rst).

Thanks a lot!

**Comment by jtsiddons** [comment]
@ludwiglierhammer: There is an element in the `icoads_r300_d730` schema named "No_data_entry" with description "Administrative number under which the data is found within the Institute". The test data for d730 has values in this field. Looking at the following elements, should this be re-named to something along the lines of "NumArchiveSet"?

See: https://github.com/glamod/cdm_reader_mapper/blob/0f8e02b00114d61067fe4b64ad81de00dd058155/cdm_reader_mapper/mdf_reader/schemas/icoads/r300/d730/icoads_r300_d730.json#L95

edit: add link to line

**Comment by jtsiddons** [comment]
Additionally mismatch between the elements "InstAbbr", "InsName", "InsPlace" etc

https://github.com/glamod/cdm_reader_mapper/blob/0f8e02b00114d61067fe4b64ad81de00dd058155/cdm_reader_mapper/mdf_reader/schemas/icoads/r300/d730/icoads_r300_d730.json#L71:L94

**Comment by ludwiglierhammer** [comment]
> @ludwiglierhammer: There is an element in the `icoads_r300_d730` schema named "No_data_entry" with description "Administrative number under which the data is found within the Institute". The test data for d730 has values in this field. Looking at the following elements, should this be re-named to something along the lines of "NumArchiveSet"?
> 
> See:
> 
> https://github.com/glamod/cdm_reader_mapper/blob/0f8e02b00114d61067fe4b64ad81de00dd058155/cdm_reader_mapper/mdf_reader/schemas/icoads/r300/d730/icoads_r300_d730.json#L95
> 
> edit: add link to line

Good point 🕵️. "NumArchiveSet" is a much better name for it, I think  .

**Comment by ludwiglierhammer** [comment]
> Additionally mismatch between the elements "InstAbbr", "InsName", "InsPlace" etc
> 
> https://github.com/glamod/cdm_reader_mapper/blob/0f8e02b00114d61067fe4b64ad81de00dd058155/cdm_reader_mapper/mdf_reader/schemas/icoads/r300/d730/icoads_r300_d730.json#L71:L94

Just re-naming "Ins" to "Inst"?

**Comment by jtsiddons** [comment]
Would you like me to fix those field names in this PR?

**Comment by ludwiglierhammer** [comment]
Maybe this is the easiest way. Then we don't need do run and fix the notebooks in two PRs.
Are you ok with this?

**Comment by jtsiddons** [comment]
Looking in more detail, there is also a mix of name formats, examples:

* CapitalCamelCase: `NameArchiveSet`
* Title_snake_case: `Logbook_language`
* camelCase: `drLatDeg`
* snake_case: `voyage_from`

Some names are abbreviated, others are complete, e.g. `drLatDeg` (Dead Reckoning degrees latitude), `Day_of_the_week`.

I've only looked at `icoads_r300_d730`.

Is it worth doing a more thorough review of the schema files for consistency?

**Comment by ludwiglierhammer** [comment]
Since those naming inconsistencies only occur in the c99 attachments ([icoads base schema](https://github.com/glamod/cdm_reader_mapper/blob/main/cdm_reader_mapper/mdf_reader/schemas/icoads/icoads.json) should be fine), we can skip the re-naming in this PR. Unless you have the time and inclination to take care of it.  If not can you please createa new issue in the [issue tracker](https://github.com/glamod/cdm_reader_mapper/issues).

**Comment by jtsiddons** [comment]
I'll fix the two specific cases outlined in this PR, and I'll file a (low priority) issue to look into naming consistency in the future.

**Comment by ludwiglierhammer** [comment]
I'll fix the column names in the testdata so that the test will pass.

**Comment by ludwiglierhammer** [comment]
@jtsiddons 
For me this PR looks good. 
Could you please add your changes to ``CHANGES.rst`` again.
After that I'd merge this PR.
Or have you planned some more changes to the notebooks?

**Comment by jtsiddons** [comment]
@ludwiglierhammer: No more changes planned for now.

Not sure which section notebook changes would fall under for `CHANGES.rst` - do you have a suggestion, or should I add a documentation section?

**Comment by ludwiglierhammer** [comment]
The idea of the change log comes from [xclim](https://github.com/Ouranosinc/xclim/blob/main/CHANGELOG.rst). They put most of the changes to the documentaion under ``New features and enhancements`` or ``Bug fixes``. In our case, I would pack it under ``Bug fixes`` and link to #103. Does that make sense in your eyes?

**Comment by jtsiddons** [comment]
Fine by me

File: CHANGES.rst [hunk 4]
```diff
@@ -78,6 +78,8 @@ Bug fixes
 * there is a loss of data precision due to data type conversion. Hence, use default data types of both ``int`` and ``float`` (:issue:`59`, :pull:`60`)
 * reading C-RAID data: adjust datetime formats to read dates into ``MDFFileReader`` (:pull:`60`)
 * ensure external code tables are used when using an external schema in ``mdf_reader.read`` (:pull:`105`)
+* restructure ``CLIWOC_datamodel`` Jupyter notebook to add an example of data model construction (:pull:`110`)
+* remove ``create_data_model.ipynb`` example Jupyter notebook (:pull:`110`)
```
Comment (ludwiglierhammer) [inline CHANGES.rst:84]: * adjust example Jupyter notebooks and README to :pull:`103` (:pull:`110`)


Breaking changes

* rename some element names for ``icoads_r300_d730`` (:pull:`110`):

  - InsName to InstName
  - InsPlace to InstPlace
  - InsLand to InstLand
  - No_data_entry to NumArchiveSet

Comment (ludwiglierhammer) [reply CHANGES.rst:84]: Could you add something like that to the CHANGES? These are all changes in this RP, aren't they?

**Comment by ludwiglierhammer** [comment]
Looks good to me. I'll merge this PR. Thanks for your help

**Pull request merged**